src/os/bluestore/BlueStore.cc

```diff
@@ -3658,41 +3658,22 @@ void BlueStore::Onode::calc_omap_tail(
   out->push_back('~');
 }
 
-void BlueStore::Onode::get() {
-  if (++nref >= 2 && !pinned) {
-    OnodeCacheShard* ocs = c->get_onode_cache();
-    ocs->lock.lock();
-    // It is possible that during waiting split_cache moved us to different OnodeCacheShard.
-    while (ocs != c->get_onode_cache()) {
-      ocs->lock.unlock();
-      ocs = c->get_onode_cache();
-      ocs->lock.lock();
-    }
-    bool was_pinned = pinned;
-    pinned = nref >= 2;
-    bool r = !was_pinned && pinned;
-    if (cached && r) {
-      ocs->_pin(this);
-    }
+void BlueStore::Onode::_pin_control()
+{
+  OnodeCacheShard* ocs = c->get_onode_cache();
+  ocs->lock.lock();
+  // It is possible that during waiting split_cache moved us to different OnodeCacheShard.
+  while (ocs != c->get_onode_cache()) {
     ocs->lock.unlock();
-  }
-}
-void BlueStore::Onode::put() {
-  ++put_nref;
-  int n = --nref;
-  if (n == 1) {
-    OnodeCacheShard* ocs = c->get_onode_cache();
+    ocs = c->get_onode_cache();
     ocs->lock.lock();
-    // It is possible that during waiting split_cache moved us to different OnodeCacheShard.
-    while (ocs != c->get_onode_cache()) {
-      ocs->lock.unlock();
-      ocs = c->get_onode_cache();
-      ocs->lock.lock();
-    }
-    bool need_unpin = pinned;
-    pinned = pinned && nref >= 2;
-    need_unpin = need_unpin && !pinned;
-    if (cached && need_unpin) {
+  }
+  bool was_pinned = pinned;
+  pinned = pin_nref >= 2;
```

Comment:
> Though pinned and pin_nref are atomic variables, this statement could be evaluated as below two: temp = pin_nref >= 2; pinned = temp;
> 
> So following case could result pin_nref = 2 but pinned = false: Suppose pin_nref = 2, pinned = true, a put thread comes in and decreases pin_nref to 1, then pin_nref >= 2 gets evaluated as false, but before it gets assigned to pinned, a get thread came in and set pinned_ref to 2 then exit because pinned is still true. After that, the put thread set the pinned 
to false and does the unpin. Finally, we'll have an unexpected state: pinned_ref=2 and pinned = false.

@taodd - hopefully I resolved this issue via moving pinning stuff to onode cache _trim_to() method (which acquires cache lock anyway). Onode ref count increment does nothing but ref count increment now.

Reply:
> The access of "pinned" variable should be restricted inside the cache lock ? I feel it's more like a member of cache shard, as its purpose is to correctly reflect the cache status of the Onode.

Using "pinned" var is replaced with calling onode lru_item's is_linked() method which explicitly indicates whether onode is in Cache LRU or not now.
And I refactored the code to isolate all the pinning/unpinning logic at cache shard rather than doing that at onode. 

So please take a look..